### PR TITLE
Fixes VSTS 915954: "Rename File to {NewClassName}.cs" code action does not work

### PR DESCRIPTION
**Pull request by avodovnik**
Adding support for the `Rename File to...` code action.

Ran into a similar issue to https://github.com/mono/monodevelop/pull/7823 and took the same workaround approach. Would make sense to discuss this, to see if there is a better (working) solution.

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs [hunk 1]
```diff
@@ -1295,8 +1296,90 @@ protected override void ApplyProjectReferenceRemoved (ProjectId projectId, Proje
...
+				SynchronizationContext.SetSynchronizationContext (oldContext);
+			}
+
+			// TODO: the Visual Studio Windows implementation of this also adds an undo unit to the
+			// global undo manager which we don't currently support.
+
+			var newName = NameGenerator.GenerateUniqueName (
+				Path.GetFileNameWithoutExtension (info.Name),
+				Path.GetExtension (info.Name),
+				nx => mdProject.Files.FirstOrDefault (p => StringComparer.OrdinalIgnoreCase.Compare (p.Name, nx) == 0) == null);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: Please use FilePath.PathComparer here. It uses the right comparison type in a cross platform manner.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs [hunk 2]
```diff
@@ -1295,8 +1296,90 @@ protected override void ApplyProjectReferenceRemoved (ProjectId projectId, Proje
...
+
+			FileService.RenameFile (guiDoc.FilePath, newName);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: Can newName be null? We should guard against that

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs (hunk 1 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: Also, shouldn't an equals check be way more efficient than comparing? The reason I'm asking is that O(n) string comparisons are expensive on big projects. Maybe we should use string.Equals (a, b, FilePath.PathComparer).

Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: We're not doing any sorting, so equals should be enough (and faster!)

Comment (avodovnik) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: Good shout, thanks! For the first pass, I was doing a naive implementation, following the [Visual Studio workspace implementation](http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/Extensions/ProjectItemsExtensions.cs,7f0b7161c3882525). :-) And you're right, we're only looking at equality, so it should be perfectly fine.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs (hunk 2 above)
Comment (avodovnik) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: I don't see how, *but* that doesn't make it impossible. I've put in a check higher up now.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs [hunk 3]
```diff
@@ -1295,8 +1296,110 @@ protected override void ApplyProjectReferenceRemoved (ProjectId projectId, Proje
...
+				mdProject = GetMonoProject (project);
+				if (mdProject == null) {
+					LoggingService.LogWarning ("Couldn't find project for newly file file {0} (Project {1}).",
+						info.Name, info.Id.ProjectId);
+
+					return;
+				}
+			}
+
+			var guiDoc = documentManager.Documents.FirstOrDefault (d => d.IsFile && d.Name == document.Name);
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: You should be comparing document.FilePath

**Comment by slluis** [comment]
Looks good. We'll need unit tests.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs (hunk 3 above)
Comment (avodovnik) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: To be honest, I don't think this is right as it is. I don't think it's trying to compare File Paths... even the "original" Windows implementation seems to be comparing names http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs,1168 

What's also weird is that the updated `DocumentInfo`'s name is actually just that, a name, without the path. I think I might need to rethink what is correct here.

Comment (slluis) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: In any case, if you want to compare the file name, you should use document.FilePath.FileName.

**Comment by avodovnik** [comment]
This was a complicated one (at least for me), so I wanted to make sure @Therzok's commit message is also noted here in the conversation:

> After debugging with @avodovnik, we found out that 2 registrations were done for the same document, one still pointing to the old path.
> 
> Subscribing to the file document controller events on initialization fixes this, as the handler that was checking renames is subscribed there.
> 
> Avoid manually saving a project, as there is a method that does that in TryApplyChanges.
> 
> The Freeze/Thaw events are required because the roslyn file associations need to happen before the text buffer registration, otherwise the text buffer is associated with the miscellaneous files workspace  as our workspace does not contain the new document file path yet.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs [hunk 4]
```diff
@@ -1295,8 +1298,124 @@ protected override void ApplyProjectReferenceRemoved (ProjectId projectId, Proje
...
+					return;
+				}
+			}
+
+			var guiDoc = documentManager
+				.Documents
+				.FirstOrDefault (d => d.IsFile
+					&& document.FilePath.Equals (d.FilePath, FilePath.PathComparison));
+
+			DispatchService.PumpingWait (() => guiDoc.IsDirty ? guiDoc.Save () : null);
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]:
```suggestion
			DispatchService.PumpingWait (() => guiDoc.IsDirty ? guiDoc.Save () : Task.CompletedTask);
```

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs [hunk 5]
```diff
@@ -1295,8 +1298,124 @@ protected override void ApplyProjectReferenceRemoved (ProjectId projectId, Proje
...
+			// global undo manager which we don't currently support.
+
+			var newName = NameGenerator.GenerateUniqueName (
+				Path.GetFileNameWithoutExtension (info.Name),
+				Path.GetExtension (info.Name),
+				nx => !mdProject.Files.Any (p => string.Equals (p.Name, nx, FilePath.PathComparison)));
+
+			var mdFile = mdProject.GetProjectFile (document.FilePath);
+			if (mdFile == null) {
+				// TODO: log this
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: Maybe just Debug.Assert?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs [hunk 6]
```diff
@@ -1295,8 +1298,124 @@ protected override void ApplyProjectReferenceRemoved (ProjectId projectId, Proje
...
+				return;
+			}
+
+			DispatchService.PumpingWait (() => {
+				return Runtime.RunInMainThread (async () => {
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]:
```suggestion
				return Runtime.RunInMainThread (() => {
```

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs [hunk 7]
```diff
@@ -1295,8 +1298,124 @@ protected override void ApplyProjectReferenceRemoved (ProjectId projectId, Proje
...
+
+					if (files == null)
+						files = new List<(MonoDevelop.Projects.ProjectFile projectFile, string name)> ();
+					files.Add ((child, childNewName));
+				}
+			}
+			return files;
+		}
+
+		void FailIfDocumentInfoChangesNotSupported (Document document, DocumentInfo updatedInfo)
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]:
```suggestion
		static void FailIfDocumentInfoChangesNotSupported (Document document, DocumentInfo updatedInfo)
```

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/TypeSystemService_WorkspaceHandling.cs [hunk 8]
```diff
@@ -310,6 +310,7 @@ public IDisposable RegisterOpenDocument (WorkspaceObject owner, FilePath filePat
 
 		void TryRegisterOpenDocument (OpenDocumentReference reference)
 		{
+ 			Console.WriteLine ($"Try Register Open Document: {reference.FilePath}");
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/TypeSystemService_WorkspaceHandling.cs]: few more writelines here.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs (hunk 5 above)
Comment (avodovnik) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: Inside the `if` statement? Is that what we're doing for this? Or is there a logging pattern?

Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: You could use logging service or Debug.Assert, when it should never happen.

Comment (avodovnik) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: It's still *possible* that it happens... :-) Let's use Logging Service, then.

**Comment by Therzok** [comment]
Still needs tests

**Comment by slluis** [comment]
@avodovnik this bug is targeting 8.2, so it should have been backported to release-8.2.

**Comment by slluis** [comment]
@monojenkins backport release-8.2

**Pull request merged**